test/e2e/network_segmentation_policy.go

```diff
@@ -0,0 +1,352 @@
+
+				ginkgo.By("asserting the *client* pod can not contact the server pod exposed endpoint")
+				gomega.Eventually(func() error {
+					return reachToServerPodFromClient(cs, serverPodConfig, clientPodConfig, serverIP, port)
+				}, 1*time.Minute, 6*time.Second).ShouldNot(gomega.Succeed())
+
+			},
+			ginkgo.Entry(
+				"in L2 dualstack primary UDN",
+				networkAttachmentConfigParams{
+					name:     nadName,
+					topology: "layer2",
+					cidr:     correctCIDRFamily(userDefinedNetworkIPv4Subnet, userDefinedNetworkIPv6Subnet),
+					role:     "primary",
+				},
+				*podConfig(
+					"client-pod",
+					withNodeSelector(map[string]string{nodeHostnameKey: workerOneNodeName}),
+				),
+				*podConfig(
+					"server-pod",
+					withCommand(func() []string {
+						return httpServerContainerCmd(port)
+					}),
+					withNodeSelector(map[string]string{nodeHostnameKey: workerTwoNodeName}),
+				),
+			),
+			ginkgo.Entry(
+				"in L3 dualstack primary UDN",
+				networkAttachmentConfigParams{
+					name:     nadName,
+					topology: "layer3",
+					cidr:     correctCIDRFamily(userDefinedNetworkIPv4Subnet, userDefinedNetworkIPv6Subnet),
+					role:     "primary",
+				},
+				*podConfig(
+					"client-pod",
+					withNodeSelector(map[string]string{nodeHostnameKey: workerOneNodeName}),
+				),
+				*podConfig(
+					"server-pod",
+					withCommand(func() []string {
+						return httpServerContainerCmd(port)
+					}),
+					withNodeSelector(map[string]string{nodeHostnameKey: workerTwoNodeName}),
+				),
+			),
+		)
+
+		ginkgo.DescribeTable(
+			"allow ingress traffic to one pod from a particular namespace",
+			func(
+				topology string,
+				clientPodConfig podConfiguration,
+				allowServerPodConfig podConfiguration,
+				denyServerPodConfig podConfiguration,
+			) {
+
+				yellow := "yellow"
+				blue := "blue"
+
+				namespaceYellow := f.Namespace.Name + "-" + yellow
+				namespaceBlue := f.Namespace.Name + "-" + blue
+
+				nad := networkAttachmentConfigParams{
+					topology: topology,
+					cidr:     correctCIDRFamily(userDefinedNetworkIPv4Subnet, userDefinedNetworkIPv6Subnet),
+					// Both yellow and blue namespaces are going to served by green network.
+					// Use random suffix for the network name to avoid race between tests.
+					networkName: fmt.Sprintf("%s-%s", "green", rand.String(randomStringLength)),
+					role:        "primary",
+				}
+				for _, namespace := range []string{namespaceYellow, namespaceBlue} {
+					ginkgo.By("Creating namespace " + namespace)
+					_, err := cs.CoreV1().Namespaces().Create(context.Background(), &v1.Namespace{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: namespace,
+						},
+					}, metav1.CreateOptions{})
+					gomega.Expect(err).NotTo(gomega.HaveOccurred())
+					defer func() {
+						gomega.Expect(cs.CoreV1().Namespaces().Delete(
+							context.Background(),
+							namespace,
+							metav1.DeleteOptions{},
+						)).To(gomega.Succeed())
+					}()
```

Comment:
Don't use defer to clean up please, it hinders troubleshooting. Use `f.AddNamespacesToDelete`; or manually in `AfterEach` depending on how the framework is configured and if the test case failed or not.

Reply:
yes @jcaamano , modified the test to use `f.AddNamespacesToDelete`.